source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,583 @@
+/*
+  Copyright (C) 2011 - 2018 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+ */
+
+
+#include <aspect/mesh_deformation/diffusion.h>
+#include <aspect/gravity_model/interface.h>
+#include <aspect/geometry_model/interface.h>
+#include <aspect/geometry_model/box.h>
+#include <aspect/geometry_model/two_merged_boxes.h>
+#include <aspect/simulator.h>
+#include <aspect/geometry_model/initial_topography_model/zero_topography.h>
+
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/base/symmetric_tensor.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <boost/lexical_cast.hpp>
+
+namespace aspect
+{
+
+  namespace MeshDeformation
+  {
+    template <int dim>
+    Diffusion<dim>::Diffusion()
+      :
+      diffusivity(0.),
+      start_time(std::numeric_limits<double>::quiet_NaN()),
+      current_time(0.),
+      last_diffusion_time(std::numeric_limits<double>::quiet_NaN()),
+      time_between_diffusion(std::numeric_limits<double>::max()),
+      start_timestep(0),
+      last_diffusion_timestep (1),
+      timesteps_between_diffusion (1),
+      apply_diffusion(false)
+    {}
+
+
+
+    template <int dim>
+    void
+    Diffusion<dim>::initialize ()
+    {
+      AssertThrow(Plugins::plugin_type_matches<GeometryModel::Box<dim> >(this->get_geometry_model()) ||
+                  Plugins::plugin_type_matches<GeometryModel::TwoMergedBoxes<dim> >(this->get_geometry_model()),
+                  ExcMessage("The surface diffusion mesh deformation plugin only works for Box geometries. "));
+
+    }
+
+
+
+    template <int dim>
+    void
+    Diffusion<dim>::update ()
+    {
+      // Initialize the start time and timestep
+      if (std::isnan(start_time))
+        {
+          start_time = (this->get_time());
+          start_timestep = this->get_timestep_number();
+        }
+
+      // Set the current time and timestep
+      current_time = (this->get_time());
+      const unsigned int current_timestep_number = this->get_timestep_number();
+
+      // Determine whether we need to apply diffusion based
+      // on the time or timestep interval between applications.
+      if (current_timestep_number != 0)
```

Comment:
Mesh velocity is solved for first thing at the beginning of each time step. At t0, mesh velocity is zero, because Stokes velocity is zero. So I didn't want to apply this mesh deformation when free surface itself isn't applied either. Also, when setting initial topography at t0, would you want it to diffuse at t0 already, or would you expect to diffuse in the following timestep?

Reply:
Ah, it's executed at the beginning of the time step. Can you add your explanation to the comment?